pkg/algorithm/cell_allocation.go

```diff
@@ -78,6 +79,77 @@ func buddyAlloc(
 	return false
 }
 
+// after buddyAlloc cannot find a healthy cell (or not in suggested nodes),
+// try to split a higher level cell safely to get current level cells
+func safeRelaxedBuddyAlloc(
+	cell *cellBindingPathVertex,
+	freeList ChainCellList,
+	freeCellNum map[CellLevel]int32,
+	currentLevel CellLevel,
+	suggestedNodes common.Set,
+	ignoreSuggestedNodes bool,
+	bindings map[api.CellAddress]*PhysicalCell) bool {
+
+	var splittableCell Cell
+	splittableNum := map[CellLevel]int32{}
+	for i := CellLevel(len(freeList)); i >= CellLevel(1); i-- {
+		// calculate splittable number
+		splittableNum[i] = int32(len(freeList[i])) - freeCellNum[i]
+		if i < CellLevel(len(freeList)) && splittableCell != nil {
+			splittableNum[i] += splittableNum[i+1] * int32(len(splittableCell.GetChildren()))
+		}
+		// iterate higher level cell
+		if splittableCell == nil && len(freeList[i]) > 0 {
+			splittableCell = freeList[i][0]
+		} else if splittableCell != nil {
+			splittableCell = splittableCell.GetChildren()[0]
+		}
+		// check safety
+		if splittableNum[i] < 0 {
+			panic(fmt.Sprintf("VC Safety Broken: level %v cell with free list %v is unsplittable, splittableNum=%v",
+				i, freeList[i], splittableNum[i]))
+		}
+	}
+
+	for l := currentLevel + 1; l <= CellLevel(len(freeList)); l++ {
+		cellNum := int32(len(freeList[l]))
+		if cellNum > splittableNum[l] {
+			cellNum = splittableNum[l]
+		}
+		if cellNum > 0 {
+			splitList := CellList{}
+			for i := int32(0); i < cellNum; i++ {
+				splitList = append(splitList, freeList[l][0])
+				freeList.remove(freeList[l][0], l)
+			}
+			splittableNum[l] -= cellNum
+			for sl := l; sl > currentLevel; sl-- {
+				for i := len(splitList); i > 0; i-- {
+					splitList = append(splitList, splitList[0].GetChildren()...)
+					copy(splitList[:], splitList[1:])
+					splitList[len(splitList)-1] = nil
```

Comment:
Why copy splitList[1:]? This is equivalent:
splitList[0] = nil
splitList = splitList[1:]